### ISSUE TITLE
Versioning guide

### ISSUE THREAD

**Issue by jonahkagan** [Clever/dev-handbook#31]
Should cover:
- why version?
- semver!!!
- bump version each time you deploy
- tips by language (e.g. for node, use `npm version`)

Questions:
- Should we dictate a convention about making separate commits for a version bump?
- Should this be part of the git guide? On one hand, it is part of the workflow. On the other hand, it has to do more with deploying than source control.

**Comment by MaerF0x0** [comment]
I think it'd be great to revive this. Frustrating to merge to master and come back after expected build time only to find NPM/Docker deploy failed due to version collisions.

**Comment by nathanleiby** [comment]
+1 to documenting this and having a process.

In order to make this more robust/automated, I'm highly in favor of Drone auto-bumping the patch version every time merge to master happens. It's safer now that we run Docker images, because consumers of the NPM package won't easily be broken by an (mistakenly) incompatible patch version.

1. until the consumer of your NPM package builds a new docker image, it won't affect deploys -- they'll keep deploying the latest image built
2. even if you've built a new docker image and the updated NPM package is for some reason incompatible, you can always deploy a prior docker image, with the older version of the NPM package

**Comment by nathanleiby** [comment]
Benefit is mitigating confusion/broken builds

**Comment by azylman** [comment]
But auto-bumping the patch version would not be following semver, which seems like a significant downside to me

**Comment by nathanleiby** [comment]
can you clarify why not following?

**Comment by nathanleiby** [comment]
Another option: change drone logic. if NPM publish fails, don't fail the build. still build docker image

**Comment by azylman** [comment]
Not all changes are patches - some are minor and some are major, so auto-patching is going to break semver

**Comment by nathanleiby** [comment]
My thought would be that eng is responsible for specifically changing the package.json for minor/major updates ONLY, before merging. 

But true, if it's automatic, someone will probably forget and eventually a minor/major change will mistakenly get automatically published in as a patch version.

**Comment by MaerF0x0** [comment]
Yeah I agree with @azylman  's sentiment against auto touching the version.

**Comment by agclever** [comment]
FWIW, I prefer doing a version bump in the PR and not post PR. I don't it should have any particular side effects but not sure.
